File: src-tauri/src/xhs.js
// Xiaohongshu (Little Red Book) site-specific DOM helpers.
// Injected into all webviews; only activates on XHS pages.
(function () {
  'use strict';

  function $(sels) {
    if (typeof sels === 'string') sels = [sels];
    for (var i = 0; i < sels.length; i++) {
      var el = document.querySelector(sels[i]);
      if (el) return el;
    }
    return null;
  }

  window.__clawXhs = {

    // ── Login check ───────────────────────────────────────────────────────
    isLoggedIn: function () {
      if ($(['.login-btn', '.login-container'])) return false;
      return !!$(['.user', '.user-info', '.sidebar-user']);
    },

    // ── Search ────────────────────────────────────────────────────────────

    setSearchInput: function (keyword) {
      var input = $(['#search-input', 'input[placeholder*="搜索"]', '.search-input input']);
      if (!input) return false;
      input.focus();
      var desc = Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value');
      if (desc && desc.set) desc.set.call(input, keyword);
      else input.value = keyword;
      input.dispatchEvent(new Event('input', { bubbles: true }));
      input.dispatchEvent(new Event('change', { bubbles: true }));
      return true;
    },

    clickSearch: function () {
      var btn = $(['.search-icon', '#search-icon', '.search-btn']);
      if (btn) { btn.click(); return true; }
      var input = $(['#search-input', 'input[placeholder*="搜索"]']);
      if (input) {
        input.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', code: 'Enter', keyCode: 13, bubbles: true }));
        return true;
      }
      return false;
    },

    // ── Feed / Results ────────────────────────────────────────────────────

    getResults: function (limit) {
      limit = limit || 20;
      var container = $(['.feeds-container', '#feeds-container']);
      if (!container) return [];
      var items = [];
      var children = Array.from(container.children);
      for (var i = 0; i < children.length && items.length < limit; i++) {
        var el = children[i];
        var isVideo = !!el.querySelector('.play-icon');
        var titleEl = el.querySelector('.title') || el.querySelector('.note-title');
        var title = titleEl ? titleEl.textContent.trim() : '';
        var cover = (el.querySelector('img') || {}).src || '';
        var authorEl = el.querySelector('.author-wrapper .name') || el.querySelector('.author .name');
        var author = authorEl ? authorEl.textContent.trim() : '';
        var likesEl = el.querySelector('.like-wrapper .count') || el.querySelector('.engagement .like .count');
        var likes = likesEl ? likesEl.textContent.trim() : '';
        if (!title && !el.textContent.trim()) continue;
        items.push({
          index: i,
          type: isVideo ? 'video' : 'note',
          title: title || el.textContent.trim().replace(/\s+/g, ' ').slice(0, 120),
          cover: cover,
          author: author,
          likes: likes
        });
      }
      return items;
    },

    // ── Note preview (overlay) ────────────────────────────────────────────

    openNote: function (index) {
      var container = $(['.feeds-container', '#feeds-container']);
      if (!container || !container.children[index]) return false;
      var target = container.children[index];
      var cover = target.querySelector('a.cover') || target.querySelector('a') || target.querySelector('img');
      if (cover) { cover.click(); return true; }
      target.click();
      return true;
    },

    closeNote: function () {
      var btn = $([
        '.close-circle', '.note-close',
        '.close', '[class*="close-btn"]'
      ]);
      if (btn) { btn.click(); return true; }
      document.dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape', keyCode: 27, bubbles: true }));
      return true;
    },

    // Whether a note overlay is currently open
    isNoteOpen: function () {
      return !!$([
        '.note-detail-mask', '.note-overlay',
        '.note-scroller', '#noteContainer'
      ]);
    },

    // ── Read note content ─────────────────────────────────────────────────

    getNoteContent: function () {
      var title = ($(['#detail-title', '.note-scroller .title', '.note-content .title']) || {}).textContent || '';
      var desc = ($(['#detail-desc', '.note-scroller .desc', '.note-content .desc', '.note-text']) || {}).innerText || '';
      var author = ($(['.author-wrapper .username', '.author-wrapper .name', '.user-nickname']) || {}).textContent || '';
      var date = ($(['.note-scroller .date', '.bottom-container .date', '.publish-date', '.date']) || {}).textContent || '';
      var likes = ($(['.interactions .like .count', '.engage-bar .like .count', '.like-wrapper .count']) || {}).textContent || '';
      var collects = ($(['.interactions .collect .count', '.engage-bar .collect .count', '.collect-wrapper .count']) || {}).textContent || '';
      var comments = ($(['.interactions .chat .count', '.engage-bar .chat .count', '.comment-wrapper .count']) || {}).textContent || '';

      var tags = [];
      var tagEls = document.querySelectorAll('#hash-tag a, .tag, .note-tag a, a[href*="search_result"]');
      for (var i = 0; i < tagEls.length; i++) {
        var t = tagEls[i].textContent.trim();
        if (t && tags.indexOf(t) === -1) tags.push(t);
      }

      var topComments = [];
      var cmtEls = document.querySelectorAll('.comment-item, .comments-container .comment, .parent-comment');
      for (var j = 0; j < cmtEls.length && j < 10; j++) {
        var cName = (cmtEls[j].querySelector('.name, .user-name, .author') || {}).textContent || '';
        var cText = (cmtEls[j].querySelector('.content, .comment-text, .text') || {}).textContent || '';
        if (cText.trim()) topComments.push({ author: cName.trim(), text: cText.trim() });
      }

      return {
        title: title.trim(),
        content: desc.trim(),
        author: author.trim(),
        date: date.trim(),
        likes: likes.trim(),
        collects: collects.trim(),
        comments: comments.trim(),
        tags: tags,
        topComments: topComments
      };
    },

    // ── Note images ───────────────────────────────────────────────────────

    getNoteImages: function () {
      var swiper = $(['.swiper-wrapper', '.carousel-wrapper', '.note-slider']);
      if (!swiper) return [];
      var imgs = swiper.querySelectorAll('img');
      var list = [];
      for (var i = 0; i < imgs.length; i++) {
        var img = imgs[i];
        var w = img.naturalWidth || img.width;
        var h = img.naturalHeight || img.height;
        if (img.src && /^https?:/.test(img.src) && w > 200 && h > 200 && list.indexOf(img.src) === -1) {
          list.push(img.src);
        }
      }
      if (list.length === 0) {
        var slides = swiper.querySelectorAll('[style*="background-image"]');
        for (var j = 0; j < slides.length; j++) {
          var m = slides[j].style.backgroundImage.match(/url\(["']?(https?:\/\/[^"')]+)["']?\)/);
          if (m && list.indexOf(m[1]) === -1) list.push(m[1]);
        }
      }
      return list;
    },

    // ── Feed scroll ───────────────────────────────────────────────────────

    scrollFeed: function () {
      window.scrollBy({ top: 800, behavior: 'smooth' });
    }
  };
})();

File: src-tauri/src/xhs.rs
//! Xiaohongshu (Little Red Book) site-specific HTTP API endpoints.
//! Mounted under `/xhs` by the main API server.

use axum::extract::{Query, State};
use axum::http::StatusCode;
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;
use tauri::AppHandle;

use crate::api::{
    eval_with_result, get_active_label, internal_err, no_active_tab, to_js_str, ErrorResponse,
    EvalResponse,
};

// ── Request types ────────────────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct SearchBody {
    pub keyword: String,
}

#[derive(Debug, Deserialize)]
pub struct OpenNoteBody {
    pub index: u32,
}

#[derive(Debug, Deserialize)]
pub struct ResultsQuery {
    #[serde(default = "default_limit")]
    pub limit: Option<u32>,
}
fn default_limit() -> Option<u32> {
    None
}

// ── Helper ───────────────────────────────────────────────────────────────────

/// Evaluate a __clawXhs method call and parse the result into a JSON value.
async fn xhs_eval(
    app: AppHandle,
    label: String,
    script: &str,
) -> Result<serde_json::Value, (StatusCode, Json<ErrorResponse>)> {
    let result_str = eval_with_result(app, label, script.to_string())
        .await
        .map_err(internal_err)?;
    let resp: EvalResponse =
        serde_json::from_str(&result_str).map_err(|e| internal_err(e.to_string()))?;
    if !resp.ok {
        return Err(internal_err(
            resp.error.unwrap_or_else(|| "xhs eval failed".to_string()),
        ));
    }
    Ok(resp.value.unwrap_or(serde_json::Value::Null))
}

// ── Handlers ─────────────────────────────────────────────────────────────────

/// GET /xhs/check-login
async fn handle_check_login(
    State(app): State<AppHandle>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let val = xhs_eval(app, label, "window.__clawXhs.isLoggedIn()").await?;
    Ok(Json(serde_json::json!({ "loggedIn": val })))
}

/// POST /xhs/search  { "keyword": "小米SU7" }
/// Sets the search input value and clicks search. Caller should /wait + /xhs/results after.
async fn handle_search(
    State(app): State<AppHandle>,
    Json(body): Json<SearchBody>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let set_script = format!(
        "window.__clawXhs.setSearchInput({})",
        to_js_str(&body.keyword)
    );
    let set_ok = xhs_eval(app.clone(), label.clone(), &set_script).await?;
    if set_ok == serde_json::Value::Bool(false) {
        return Err((
            StatusCode::BAD_REQUEST,
            Json(ErrorResponse {
                error: "Search input not found — is this a Xiaohongshu page?".to_string(),
            }),
        ));
    }
    // Small delay then click search
    tokio::time::sleep(std::time::Duration::from_millis(300)).await;
    let click_ok = xhs_eval(app, label, "window.__clawXhs.clickSearch()").await?;
    Ok(Json(
        serde_json::json!({ "inputSet": set_ok, "searchClicked": click_ok }),
    ))
}

/// GET /xhs/results?limit=20
async fn handle_results(
    State(app): State<AppHandle>,
    Query(params): Query<ResultsQuery>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let limit = params.limit.unwrap_or(20);
    let script = format!("window.__clawXhs.getResults({})", limit);
    let items = xhs_eval(app, label, &script).await?;
    Ok(Json(serde_json::json!({ "items": items })))
}

/// POST /xhs/open-note  { "index": 0 }
/// Clicks the note at the given feed index to open the overlay.
/// Caller should wait ~1.5 s then call /xhs/note-content.
async fn handle_open_note(
    State(app): State<AppHandle>,
    Json(body): Json<OpenNoteBody>,
) -> Result<StatusCode, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let script = format!("window.__clawXhs.openNote({})", body.index);
    let ok = xhs_eval(app, label, &script).await?;
    if ok == serde_json::Value::Bool(false) {
        return Err((
            StatusCode::BAD_REQUEST,
            Json(ErrorResponse {
                error: format!("Note at index {} not found", body.index),
            }),
        ));
    }
    Ok(StatusCode::NO_CONTENT)
}

/// POST /xhs/close-note
async fn handle_close_note(
    State(app): State<AppHandle>,
) -> Result<StatusCode, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    xhs_eval(app, label, "window.__clawXhs.closeNote()").await?;
    Ok(StatusCode::NO_CONTENT)
}

/// GET /xhs/note-content
/// Reads the currently open note's title, content, author, stats, tags, and top comments.
async fn handle_note_content(
    State(app): State<AppHandle>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let note = xhs_eval(app, label, "window.__clawXhs.getNoteContent()").await?;
    Ok(Json(note))
}

/// GET /xhs/note-images
/// Returns image URLs from the current note's swiper/carousel.
async fn handle_note_images(
    State(app): State<AppHandle>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let images = xhs_eval(app, label, "window.__clawXhs.getNoteImages()").await?;
    Ok(Json(serde_json::json!({ "images": images })))
}

/// POST /xhs/scroll-feed
/// Scrolls down 800px to trigger lazy loading of more results.
async fn handle_scroll_feed(
    State(app): State<AppHandle>,
) -> Result<StatusCode, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    xhs_eval(app, label, "window.__clawXhs.scrollFeed()").await?;
    Ok(StatusCode::NO_CONTENT)
}

/// GET /xhs/is-note-open
async fn handle_is_note_open(
    State(app): State<AppHandle>,
) -> Result<Json<serde_json::Value>, (StatusCode, Json<ErrorResponse>)> {
    let label = get_active_label(&app).ok_or_else(no_active_tab)?;
    let open = xhs_eval(app, label, "window.__clawXhs.isNoteOpen()").await?;
    Ok(Json(serde_json::json!({ "open": open })))
}

// ── Router ───────────────────────────────────────────────────────────────────

pub fn xhs_router() -> Router<AppHandle> {
    Router::new()
        .route("/check-login", get(handle_check_login))
        .route("/search", post(handle_search))
        .route("/results", get(handle_results))
        .route("/open-note", post(handle_open_note))
        .route("/close-note", post(handle_close_note))
        .route("/note-content", get(handle_note_content))
        .route("/note-images", get(handle_note_images))
        .route("/scroll-feed", post(handle_scroll_feed))
        .route("/is-note-open", get(handle_is_note_open))
}
